x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -0,0 +1,122 @@
+
+import { isEmpty, flatten } from 'lodash';
+import { callClusterFactory } from '../../../xpack_main';
+import {
+  getCollectorLogger,
+  sendBulkPayload,
+  monitoringBulk,
+} from './lib';
+
+/*
+ * Handles internal Kibana stats collection and uploading data to Monitoring
+ * bulk endpoint.
+ *
+ * NOTE: internal collection will be removed in 7.0
+ *
+ * Depends on
+ *   - 'xpack.monitoring.kibana.collection.enabled' config
+ *   - monitoring feature from xpack_main.info isAvailable() and isEnabled()
+ * The dependencies are handled upstream
+ * - Ops Events - essentially Kibana's /api/status
+ * - Usage Stats - essentially Kibana's /api/stats
+ * - Kibana Settings - select uiSettings
+ * @param {Object} server HapiJS server instance
+ * @param {Object} xpackInfo server.plugins.xpack_main.info object
+ */
+export class BulkUploader {
+  constructor(server, xpackMainInfo, { interval, combineTypes }) {
+    if (typeof interval !== 'number') {
+      throw new Error('interval number of milliseconds is required');
+    }
+    if (typeof combineTypes !== 'function') {
+      throw new Error('combineTypes function is required');
+    }
+
+    this._timer =  null;
+    this._xpackMainInfo = xpackMainInfo;
+    this._interval = interval;
+    this._combineTypes = combineTypes;
+    this._log = getCollectorLogger(server);
+
+    this._client = server.plugins.elasticsearch.getCluster('admin').createClient({
+      plugins: [monitoringBulk],
+    });
+
+    this._callClusterWithInternalUser = callClusterFactory(server).getCallClusterInternal();
+
+  }
+
+  /*
+   * Start the interval timer
+   * @param {CollectorSet} collectorSet object to use for initial the fetch/upload and fetch/uploading on interval
+   * @return undefined
+   */
+  start(collectorSet) {
+    this._log.info('Starting monitoring stats collection');
+    if (this._timer) {
+      this._log.warn('BulkUploader timer already started');
+      return;
+    }
+
+    this._fetchAndUpload(collectorSet); // initial fetch
+    this._timer = setInterval(() => {
+      this._fetchAndUpload(collectorSet);
+    }, this._interval);
+  }
+
+  /*
+   * start() and stop() are lifecycle event handlers for
+   * xpackMainPlugin state changes
+   */
+  stop() {
+    this._log.info('Stopping monitoring stats collection');
+    clearInterval(this._timer);
+    this._timer = null;
+  }
+
+  /*
+   * @param {CollectorSet} collectorSet
+   * @return {Promise} - resolves to undefined
+   */
+  async _fetchAndUpload(collectorSet) {
+    // Before every fetch, check the license and make sure the bulk endpoint on the ES side is up
+    const mainMonitoring = this._xpackMainInfo.feature('monitoring');
+    const monitoringBulkEnabled = mainMonitoring && mainMonitoring.isAvailable() && mainMonitoring.isEnabled();
+
+    if (!monitoringBulkEnabled) {
+      // debug level to not spam the logs
+      this._log.debug(`Skipping fetch and upload of monitoring stats due to monitoring bulk endpoint not being available`);
```

Comment:
Continuing to check at an interval protects the possibility that the bulk endpoint in ES can be disabled with a dynamic setting (not sure if that is possible), or that the ES configuration changes without triggering an ES plugin state change in Kibana. That seems like it is possible with hot-swapping ES for one with a different configuration.

Reply:
@tsullivan The endpoint will not change based on the license, nor can it be changed dynamically (the endpoint either exists, or it doesn't because `xpack.monitoring.enabled: false` is explicitly set). However, calling `this.stop()` would require the `xpack_main` plugin to go from green to another state, then back to green for the service to be restarted, which is _not_ a requirement for a license change.

Relative to the next comment, I think that is the superior option and then we can add this logic there.